### PR TITLE
Remove mask_all_nan kwarg and handle input fill_values in average bucket resampler, introduce skipna in get_average and get_sum

### LINKED ISSUE #317: Fix mask_all_nan kwarg in Average Bucket Resampler
```python
from pyresample import create_area_def
from pyresample.bucket import BucketResampler
import dask.array as da
import numpy as np

area = create_area_def(
    area_id='test',
    projection={'proj': 'latlong'},
    center=(0., 0.),
    width=1,
    height=1,
    resolution=10.0
)

lons = da.from_array(np.array([0, 0, 0, 0]), chunks=2)
lats = da.from_array(np.array([0, 0, 0, 0]), chunks=2)
resampler = BucketResampler(source_lats=lats,
                            source_lons=lons,
                            target_area=area)


print('No fill_value passed:')

data = da.from_array(np.array([1, 2, 3, 4]), chunks=2)
print('mask_all_nan=False, all valid values: ', resampler.get_average(data, mask_all_nan=False).compute())
print('mask_all_nan=True,  all valid values: ', resampler.get_average(data, mask_all_nan=True).compute())

data = da.from_array(np.array([1, 2, np.nan, 4]), chunks=2)
print('mask_all_nan=False, one NaN value: ', resampler.get_average(data, mask_all_nan=False).compute())
print('mask_all_nan=True,  one NaN value: ', resampler.get_average(data, mask_all_nan=True).compute())

data = da.from_array(np.array([np.nan, np.nan, np.nan, np.nan]), chunks=2)
print('mask_all_nan=False, all NaN values: ', resampler.get_average(data, mask_all_nan=False).compute())
print('mask_all_nan=True,  all NaN values: ', resampler.get_average(data, mask_all_nan=True).compute())

print('\nfill_value passed as -999:')

data = da.from_array(np.array([1, 2, 3, 4]), chunks=2)
print('mask_all_nan=False, all valid values: ', resampler.get_average(data, fill_value=-999, mask_all_nan=False).compute())
print('mask_all_nan=True,  all valid values: ', resampler.get_average(data, fill_value=-999, mask_all_nan=True).compute())

data = da.from_array(np.array([1, 2, np.nan, 4]), chunks=2)
print('mask_all_nan=False, one NaN value: ', resampler.get_average(data, fill_value=-999, mask_all_nan=False).compute())
print('mask_all_nan=True,  one NaN value: ', resampler.get_average(data, fill_value=-999, mask_all_nan=True).compute())

data = da.from_array(np.array([np.nan, np.nan, np.nan, np.nan]), chunks=2)
print('mask_all_nan=False, all NaN values: ', resampler.get_average(data, fill_value=-999, mask_all_nan=False).compute())
print('mask_all_nan=True,  all NaN values: ', resampler.get_average(data, fill_value=-999, mask_all_nan=True).compute())


```
#### Problem description
The kwarg `mask_all_nan` has no influence on the results, no matter how many NaNs are present in the data, and if the `fill_value` is set or not. As it is supposed to work now, it doesn't make much sense in my opinion anyway, since the result of the average of all-NaN can only be NaN/FillValue anyway.

I would propose to repurpose the `mask_all_nan` kwarg such that, if set `True`, the averaged bucket is NaN if at least one of the elements in the bucket is NaN. If the kwarg is `False`, the average should be computed from the valid elements (as it's done now, i.e. a `nanmean` functionality). 

Or, even better, should we get rid of the old `mask_all_nan` kwarg and introduce a new one, e.g. called `mask_if_one_nan`, for this functionality? This would probably be the most clean and transparent thing to do, I would say. Feedback appreciated :)

#### Expected Output
```bash
No fill_value passed:
mask_all_nan=False, all valid values:  [[2.5]]
mask_all_nan=True,  all valid values:  [[2.5]]
mask_all_nan=False, one NaN value:  [[2.33333333]]
mask_all_nan=True,  one NaN value:  [[nan]]
mask_all_nan=False, all NaN values:  [[nan]]
mask_all_nan=True,  all NaN values:  [[nan]]

fill_value passed as -999:
mask_all_nan=False, all valid values:  [[2.5]]
mask_all_nan=True,  all valid values:  [[2.5]]
mask_all_nan=False, one NaN value:  [[2.33333333]]
mask_all_nan=True,  one NaN value:  [[-999]]
mask_all_nan=False, all NaN values:  [[-999.]]
mask_all_nan=True,  all NaN values:  [[-999.]]
```

#### Actual Result
```bash
No fill_value passed:
mask_all_nan=False, all valid values:  [[2.5]]
mask_all_nan=True,  all valid values:  [[2.5]]
mask_all_nan=False, one NaN value:  [[2.33333333]]
mask_all_nan=True,  one NaN value:  [[2.33333333]]
mask_all_nan=False, all NaN values:  [[nan]]
mask_all_nan=True,  all NaN values:  [[nan]]

fill_value passed as -999:
mask_all_nan=False, all valid values:  [[2.5]]
mask_all_nan=True,  all valid values:  [[2.5]]
mask_all_nan=False, one NaN value:  [[2.33333333]]
mask_all_nan=True,  one NaN value:  [[2.33333333]]
mask_all_nan=False, all NaN values:  [[-999.]]
mask_all_nan=True,  all NaN values:  [[-999.]]
```

### PR DESCRIPTION
**Pull request by ameraner**
- [x] Closes #317  
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``git diff origin/master **/*py | flake8 --diff`` 
 - [x] Fully documented

### CODE REVIEW

**Comment by ameraner** [comment]
@pnuu @sfinkens PR is ready as discussed in linked issue!

**Review by sfinkens — COMMENTED** [review]
Nice work, looks good! One idea to maybe save some `isnan` and `where` calls: Would it be possible to pass the `skipna` kwarg down to `get_sum` and, if False, replace this line

https://github.com/pytroll/pyresample/blob/bc1cdc7e072b11bf3b7a8f813bb95546e708a196/pyresample/bucket/__init__.py#L160-L161

with 
```
weights = data
```
Would that have the same effect like the `if skipna` block in your current approach?

**Comment by ameraner** [comment]
@sfinkens good idea indeed! Didn't see your edited comment but I had the same solution :)
If this looks good, I'll add tests for `get_sum` tomorrow. 

We should probably add `fill_value` to `get_sum` as well, right? And in that case keeping the input dtype would make sense, I'd say. But I guess this is material for another PR.

**Comment by sfinkens** [comment]
Looks good to me! Thanks for the detailed docstrings :+1:

**Comment by ameraner** [comment]
I implemented a workaround for the dask histogram bug (see https://github.com/dask/dask/issues/6935). The current behavior is as discussed in slack, i.e. following the numpy nan/nanmean/sum/nansum behavior:
```
get_average:
skipna=False -> bucket with one NaN or more to be NaN
skipna=True  -> bucket is NaN for all-NaN buckets or result of product
empty buckets are nan in both cases

get_sum:
skipna=False -> bucket with one NaN or more to be NaN
skipna=True  -> bucket is zero for all-NaN buckets or result of sum
empty buckets are 0 in both cases
```

If there are no further comments, I would propose to merge this PR, as it achieved the original purpose. Once the dask bug is fixed, the workaround can be removed (I'll create an issue for it). In subsequent PRs, we can add the discussed `mask_empty_buckets` and `mask_allnan_buckets`.

**Review by sfinkens — APPROVED** [review]
LGTM, nice job! I think the tests would benefit from a smaller target area, then we could compare the entire output arrays more easily. But that's also something for another PR :slightly_smiling_face:

File: pyresample/bucket/__init__.py [hunk 1]
```diff
@@ -169,12 +176,12 @@ def get_sum(self, data, mask_all_nan=False):
         sums, _ = da.histogram(self.idxs, bins=out_size, range=(0, out_size),
                                weights=weights, density=False)
 
-        if mask_all_nan:
+        # TODO remove following lines in favour of weights = data when dask histogram bug (issue #6935) is fixed
```
Comment (mraspaud) [inline pyresample/bucket/__init__.py:179]: Can you make a separate method for the lines that would have to be removed when the bug is fixed?

File: pyresample/test/test_bucket.py [hunk 2]
```diff
@@ -107,44 +106,44 @@ def test_get_bucket_indices(self):
...
+    def _get_sum_result(self, data, **kwargs):
+        """Compute the bucket average with kwargs and check that no dask computation is performed."""
+        with dask.config.set(scheduler=CustomScheduler(max_computes=0)):
+            result = self.resampler.get_sum(data, **kwargs)
+        return result.compute()
+
     def test_get_sum(self):
         """Test drop-in-a-bucket sum."""
-        data = da.from_array(np.array([[2., 2.], [2., 2.]]),
+        data = da.from_array(np.array([[2., 3.], [7., 16.]]),
```
Comment (mraspaud) [inline pyresample/test/test_bucket.py:117]: I would split `test_get_sum` into multiple test functions, what do you think?

**Review by mraspaud — COMMENTED** [review]
LGTM, a couple minor styling comments.

**Review by pnuu — APPROVED** [review]
LGTM

File: pyresample/bucket/__init__.py (hunk 1 above)
Comment (ameraner) [reply pyresample/bucket/__init__.py]: Done in a new commit, is it ok like this?

Comment (mraspaud) [reply pyresample/bucket/__init__.py]: Thanks, looks good

File: pyresample/test/test_bucket.py (hunk 2 above)
Comment (ameraner) [reply pyresample/test/test_bucket.py:117]: Good idea! Just added a commit, I also split the tests for `get_average`

Comment (mraspaud) [reply pyresample/test/test_bucket.py:117]: Looks great!

**Review by mraspaud — APPROVED** [review]
LGTM!

**Comment by ameraner** [comment]
Thanks @sfinkens @pnuu @mraspaud for your help on this!

**Pull request merged**